xarray/namedarray/core.py

```diff
@@ -844,5 +853,141 @@ def _to_dense(self) -> NamedArray[Any, _DType_co]:
+
+        if not dims:
+            dims = self.dims[::-1]
+        else:
+            dims = tuple(infix_dims(dims, self.dims, missing_dims))
+
+        if len(dims) < 2 or dims == self.dims:
+            # no need to transpose if only one dimension
+            # or dims are in same order
+            return self.copy(deep=False)
+
+        axes = self.get_axis_num(dims)
+        data = as_indexable(self._data).transpose(axes)  # type: ignore
+        return self._replace(dims=dims, data=data)
+
+    @property
+    def T(self) -> NamedArray[Any, _DType_co]:
+        return self.permute_dims()
+
+    def _check_dims(self, dims: Mapping[Any, _Dim] | _Dim):
+        if isinstance(dims, dict):
+            dims_keys = dims
+        else:
+            dims_keys = dims if isinstance(dims, str) else list(dims)
+        if missing_dims := set(self.dims) - set(dims_keys):
+            raise ValueError(
+                f"new dimensions {dims!r} must be a superset of "
+                f"existing dimensions {self.dims!r}. missing dims: {missing_dims}"
+            )
+
+    def _get_expanded_dims(self, dims: Mapping[Any, _Dim] | _Dim) -> _DimsLike:
+        self_dims = set(self.dims)
+        return tuple(dim for dim in dims if dim not in self_dims) + self.dims
+
+    def broadcast_to(self, dims: Mapping[Any, _Dim]) -> NamedArray[Any, _DType_co]:
+        """
+        Broadcast namedarray to a new shape.
+
+        Parameters
+        ----------
+        dims : dict
+            Dimensions to broadcast the array to. Keys are dimension names and values are the new sizes.
+        """
+
+        from xarray.core import duck_array_ops  # TODO: Remove this import in the future
+
+        self._check_dims(dims)
+        expanded_dims = self._get_expanded_dims(dims)
+        shape = list(dims.values())
+
+        dims_map = dict(zip(dims, cast(Iterable[SupportsIndex], shape)))
+        temporary_shape = tuple(dims_map[dim] for dim in expanded_dims)
+
+        data = duck_array_ops.broadcast_to(self.data, temporary_shape)  # type: ignore
+        return self._new(data=data, dims=expanded_dims)
+
+    def expand_dims(
+        self, dims: _DimsLike, shape: _ShapeLike | None = None
+    ) -> NamedArray[Any, _DType_co]:
+        """
+        Expand the dimensions of the object.
+
+        This method adds new dimensions to the object and optionally broadcasts
+        the data to the new shape if provided.
+
+        Parameters
+        ----------
+        dims : str or sequence of str or dict
+            Dimensions to include on the new object (must be a superset of the existing dimensions).
+            If a dict, values are used to provide the sizes of new dimensions; otherwise, new dimensions are inserted with length 1.
+
+        shape : sequence of int, optional
+            Shape to broadcast the data to. Must be specified in the same order as `dims`.
+            If not provided, new dimensions are inserted with length 1.
```

Comment:
```suggestion
```

I think `dims: str | Sequence[str] | Mapping, **dims_kwargs` is the right signature here. This forces correctness on the user.

There is a use case for adding the `axis` kwarg to control where the new dims are inserted https://docs.xarray.dev/en/stable/generated/xarray.DataArray.expand_dims.html . But it'd be nice to stick it in the Mapping too if we can figure out a nice syntax.

Reply:
Or perhaps this should be `expand_dims({dim_name: axis_number})` inserting size 1 always. And we ask the user to use `broadcast_to` to broadcast out.